Allow fasttail calls for engregesterable structs

jashook opened the pull request:
**Note**

This causes 153 new fastTailCall decisions in System.Private.CoreLib. With no new decisions to not fastTailCall. See [this list for new fastTailCalled methods](https://gist.github.com/jashook/39fe74a51520a13f3c6e36282612ae14).

Commit:
[*64] Allow more fastTailCalls involving structs

Before this change structs on Arm64 and Amd64 Unix could
pessimize when we could fastTailCall if they were engregisterable
and took more than one register.

briansull commented on src/jit/lclvars.cpp:
```diff
@@ -233,16 +240,55 @@ void Compiler::lvaInitTypeRef()
+        varDsc->IsFloatRegType() ? ++floatingRegCount : ++argRegCount;
+    };
+
+    for (unsigned argNum = 0; argNum < info.compMethodInfo->args.numArgs; argNum++, varDscTrailingPointer++)
```
Consider this version of this loop used elsewhere:

`for (lclNum = 0, varDsc = lvaTable; lclNum < lvaCount; lclNum++, varDsc++)`

briansull commented on src/jit/lclvars.cpp:
```diff
@@ -233,16 +240,55 @@ void Compiler::lvaInitTypeRef()
     varDscInfo.Init(lvaTable, hasRetBuffArg);
 
+    // Maintain a pointer to the start of the arguments.
+    LclVarDsc* varDscTrailingPointer = varDscInfo.varDsc;
```
renames as curArg or similiar

briansull commented on src/jit/lclvars.cpp:
```diff
@@ -233,16 +240,55 @@ void Compiler::lvaInitTypeRef()
     CORINFO_ARG_LIST_HANDLE localsSig = info.compMethodInfo->locals.args;
 
+    // Maintain a pointer to the start of the locals.
+    varDscTrailingPointer = varDsc;
```
Is this a dead assignment?

briansull commented on src/jit/lclvars.cpp:
```diff
@@ -255,13 +301,25 @@ void Compiler::lvaInitTypeRef()
         varDsc->lvOnFrame = true; // The final home for this local variable might be our local stack frame
 
+        CORINFO_CLASS_HANDLE clsHnd = info.compCompHnd->getArgClass(&info.compMethodInfo->locals, localsSig);
+        stackSize += roundUp(compGetTypeSize(strip(corInfoType), clsHnd), TARGET_POINTER_SIZE);
```
This local name is confusing:  `corInfoType `

briansull commented on src/jit/morph.cpp:
```diff
@@ -6943,20 +6943,33 @@ bool Compiler::fgCanFastTailCall(GenTreeCall* callee)
+    unsigned callerArgRegCount        = info.compArgRegCount;
+    unsigned callerFloatArgRegCount   = info.compFloatArgRegCount;
+
+    // TODO ARM64, UNIX x64
```
Need `// TODO-ARM64/AMD64-UNIX` instead

briansull commented on src/jit/morph.cpp:
```diff
@@ -7001,24 +7012,73 @@ bool Compiler::fgCanFastTailCall(GenTreeCall* callee)
+                //
+                // hasMultiByteArgs will determine if the struct can be passed
+                // in registers. If it cannot we will break the loop and not
+                // fastTailCall.
```
I would prefer to have each line end with a period,  NOTE is implied here

briansull commented on src/jit/morph.cpp:
```diff
@@ -7001,24 +7012,73 @@ bool Compiler::fgCanFastTailCall(GenTreeCall* callee)
 
-                if (typeSize > TARGET_POINTER_SIZE)
+                // TODO. Here we have made the assumption that multibyte struct
+                // arguments will cause a no fastTailCall decision.
```
// TODO-AMD64-Linux

BruceForstall commented:
In general, it would be useful to have a file-level comment for your new test cases, describing what they are intending to test -- an overall "here's why this is here" kind of comment.

jashook commented:
Ok this Pr is in a good state again. It has changes based on @briansull's comments that are incomplete.

BruceForstall commented on src/jit/lclvars.cpp:
```diff
@@ -108,12 +108,19 @@ void Compiler::lvaInitTypeRef()
+        ++info.compArgsCount;
+        ++argRegCount;
+    };
 
```
This code assumes that anything using incrementArgCount (which is the four specials: 'this', retbuf, generics context, varargs cookie) are register arguments. But that doesn't appear to be true, e.g., for x86, which only has 2 register arguments. (There are also "non-standard" reg args for things like pinvoke helpers, etc.).

It seems like the only code that can count reg args correctly is fgMorphArgs, when AddRegArg() is called. I wonder if this calculation should be done then instead.

BruceForstall replied:
Ok, ignore that; I was confusing incoming and outgoing args.

BruceForstall commented on src/jit/lclvars.cpp:
```diff
@@ -255,13 +301,25 @@ void Compiler::lvaInitTypeRef()
 
+        CORINFO_CLASS_HANDLE clsHnd = info.compCompHnd->getArgClass(&info.compMethodInfo->locals, localsSig);
+        stackSize += roundUp(compGetTypeSize(strip(corInfoType), clsHnd), TARGET_POINTER_SIZE);
+
```
You hoisted this out of the ` if (strip(corInfoType) == CORINFO_TYPE_CLASS)` condition. Is it legal (or desirable) to call `getArgClass` in all cases?

Commit:
Exclude stackBound two slot size structs from fastTailCalls

BruceForstall commented on src/jit/lclvars.cpp:
```diff
@@ -108,12 +108,19 @@ void Compiler::lvaInitTypeRef()
+        ++info.compArgsCount;
+        ++argRegCount;
+    };
 
```
Isn't this going to double-count the "specials"? Don't they have varDsc that is going to be counted by your new loop below the call to lvaInitArgs()?

BruceForstall commented on src/jit/lclvars.cpp:
```diff
@@ -236,6 +243,40 @@ void Compiler::lvaInitTypeRef()
+            incrementRegCount(varDsc);
+
+#ifdef (FEATURE_MULTIREG_ARGS && !WINDOWS_AMD64_ABI)
+            if (varDsc->lvOtherArgReg != REG_NA)
```
Doesn't this need to be written:
```
#if defined(FEATURE_MULTIREG_ARGS) && !defined(WINDOWS_AMD64_ABI)
```

BruceForstall replied:
Instead of `!defined(WINDOWS_AMD64_ABI)` I would prefer `defined(UNIX_AMD64_ABI)`

---
In reply to: [124107782](https://github.com/jashook/coreclr/pull/6#discussion_r124107782) [](ancestors = 124107782)

BruceForstall commented on src/jit/lclvars.cpp:
```diff
@@ -236,6 +243,40 @@ void Compiler::lvaInitTypeRef()
+            incrementRegCount(varDsc);
+
+#ifdef (FEATURE_MULTIREG_ARGS && !WINDOWS_AMD64_ABI)
+            if (varDsc->lvOtherArgReg != REG_NA)
```
although it's not clear it needs any platform `#ifdef`. If it does, please add a comment as to why.

---
In reply to: [124108646](https://github.com/jashook/coreclr/pull/6#discussion_r124108646) [](ancestors = 124108646,124107782)

BruceForstall commented on src/jit/lclvars.cpp:
```diff
@@ -236,6 +243,40 @@ void Compiler::lvaInitTypeRef()
+    // Calculate the argument register usage.
+    //
+    // This will later be used for fastTailCall determination
+    //-------------------------------------------------------------------------
```
This whole section should probably be under `#if FEATURE_FASTTAILCALL`

BruceForstall commented on src/jit/lclvars.cpp:
```diff
@@ -236,6 +243,40 @@ void Compiler::lvaInitTypeRef()
+    for (LclVarDsc* varDsc, unsigned argNum = 0; argNum < info.compMethodInfo->args.numArgs; argNum++, varDsc++)
+    {
+        if (varDsc->lvRegNum != REG_STK)
+        {
```
shouldn't this be lvArgReg? or maybe lvIsRegArg?

BruceForstall commented on src/jit/lclvars.cpp:
```diff
@@ -236,6 +243,40 @@ void Compiler::lvaInitTypeRef()
+        else
+        {
+            stackSize += varDsc->lvSize();
+        }
```
lvSize() can only be used for structs. I think you need a different calculation here for argument stack size.

BruceForstall commented on src/jit/morph.cpp:
```diff
@@ -7001,24 +7013,83 @@ bool Compiler::fgCanFastTailCall(GenTreeCall* callee)
+                }
+
+#else // ARM64
+                var_types hfaType  = GetHfaType(argx);
```
also Windows amd64, which I think you don't want?

BruceForstall commented on src/jit/morph.cpp:
```diff
@@ -7038,20 +7113,75 @@ bool Compiler::fgCanFastTailCall(GenTreeCall* callee)
+    }
+
+#elif (_TARGET_AMD64_ && UNIX_AMD64_ABI) || _TARGET_ARM64_
+
```
I think this needs to be (or, at least, should be) written:
```
#elif (defined(_TARGET_AMD64_) && defined(UNIX_AMD64_ABI)) || defined(_TARGET_ARM64_)

BruceForstall commented on src/jit/morph.cpp:
```diff
@@ -7038,20 +7113,75 @@ bool Compiler::fgCanFastTailCall(GenTreeCall* callee)
+
+    // For *nix Amd64 and Arm64 check to see if all arguments for the callee
+    // and caller are passing in registers. If not make sure that the stack
+    // size for the callee is at less than or equal to the caller's.
```
>is at less [](start = 27, length = 10)

> "... is at less ..."

Should the "at" be removed? I might rephrase to say:

```
If not, ensure that the outgoing argument stack size requirement for the callee is less than or equal to the caller's entire stack frame usage.
```

(I assume we care about the caller's entire stack frame usage, not the caller's argument stack space usage, which would be less.)

BruceForstall commented on src/jit/morph.cpp:
```diff
@@ -7038,20 +7113,75 @@ bool Compiler::fgCanFastTailCall(GenTreeCall* callee)
+    // x64 Windows: If we have more callee registers used than MAX_REG_ARG, then 
+    // make sure the callee's incoming arguments is less than the caller's
+    if ((((calleeArgRegCount + calleeFloatArgRegCount) > maxRegArgs) && 
+        ((callerArgRegCount + callerFloatArgRegCount) > (callerArgRegCount + callerFloatArgRegCount)))
```
>(callerArgRegCount + callerFloatArgRegCount) > (callerArgRegCount + callerFloatArgRegCount) [](start = 9, length = 91)

There's a problem here:
```
(callerArgRegCount + callerFloatArgRegCount) > (callerArgRegCount + callerFloatArgRegCount)
```
looks like `a > a`

Commit:
Address feedback

Commit:
Fix a > a

Commit:
Fix log message

Commit:
Requiring nCalleeArgs==nCallerArgs if hasStackArgs

LowerFastTailCall assumes nCallerArgs == nCalleeArgs. This
is tracked with https://github.com/dotnet/coreclr/issues/12468.
It should be fixed in a seperate PR.

Commit:
Remove debug code.

Commit:
Fix windows build breaks

Commit:
Fix windows warnings

Commit:
Fix last windows warning

Commit:
Fix hopefully the last windows warning

Commit:
Make the correct fastTailCall decision on Windows

Now shows no diff in fastTailCall decisions on Amd64 Windows.

briansull commented on src/jit/morph.cpp:
```diff
@@ -7001,24 +7018,85 @@ bool Compiler::fgCanFastTailCall(GenTreeCall* callee)
-#endif // FEATURE_UNIX_AMD64_STRUCT_PASSING || _TARGET_ARM64_
+                else
+                {
+                    if (structDesc.eightByteCount > 1)
```
This reads better as:

`if (structDesc.eightByteCount == 2)`

briansull reviewed (commented):
Approved with one comment
